### DISCUSSION TITLE
More realistic road speed

### DISCUSSION THREAD

**Discussion by Esmolante** [organicmaps/organicmaps#3072] [Ideas]
Like all other navigators, the algorithm for calculating the fastest route assumes that the driver will travel at the maximum speed of the road. This assumption can be misleading, and provide a sub-optimal route calculation. 
An example: There are two ways to reach the destination, one on a national road (maxspeed=100 km/h), and a slightly shorter way through a mountain pass (maxspeed=90 km/h). But in reality, the average speed on the latter road is 50 km/h. No one who knows the area would choose the latter option, as it is slower, although the calculation using the maximum legal speed on the road indicates otherwise. And even if it were slightly faster, most drivers would still prefer the better road, as it is the more comfortable, safer and more fuel efficient alternative.
So what do I propose to improve the route calculation? Use in the algorithm, not maxspeed, but a more realistic average speed, defined on the basis of road geometry and gradient. Of course, this would be a parameter specific to each road section that would have to be calculated in advance, when importing the OSM maps. And it would be sufficient for the algorithm to work with this speed value, instead of maxspeed.
Similarly, one could think about how to calculate a speed in a city, because it is not reasonable to think that we can cross a big city at 50 km/h maintained. There is dense traffic, traffic lights, crossings, pedestrian crossings... Here it is not very clear to me how to do it. It could be modulated by time of day (at night closer to maxspeed), and add delay for each traffic light, junction, etc. All the navigators, with the exception of Google Maps (it plays in another league, as it has access to real position data), suggest me to go through Madrid instead of going around it by the ring roads, which is crazy.
The proposed changes could be the differentiating factor to make OM the best offline browser. Or, at least, the best at calculating routes, which for me is almost a synonym.

**Comment by biodranik** [comment]
You're mixing it all together :)
1. Access to traffic data can improve any city/important highway navigation. It is impossible to do without the live traffic collection or at least statistical models in the past.
2. Introducing turns and traffic lights penalties should improve the algo.
3. Alternate routes should help in your 100/90km examples.

**Comment by Esmolante** [comment]
But alternative routes are of no use when the driver does not know the route (which is what the navigator is for). The driver chooses the most direct route through the mountain pass, unaware of the trouble he is getting into until he is already there, cursing the navigator for taking him that way.
Therefore, my proposal would be, when building the map for OM from OSM data, to analyse all roads from the points that define their axis and associate them with an average speed predicted by their geometry and gradient.
One way of doing this would be to take a point and calculate the distance to the straight line formed by the two previous points. And relate this value to the distance to the previous point. In this way we can discriminate whether it is an open or closed curve. By running this function over the whole layout of each section of road, we can get a faithful idea of its geometry (if it has many straight lines, open curves, closed curves...). With these values we would define a parameter which we could call "curvicity", which would limit the maxspeed of the road. It would be much simpler to do the same, but for the accumulated gradient, as steep gradients also slow down the pace.
The important thing is to approximate each stretch of road to the real speed at which an average driver would drive along it. This would allow the route calculation algorithm to work with much more realistic values than using the maxspeed value, offering more refined routes and more realistic estimated arrival times.

**Reply by biodranik** [reply]
CC @vng

**Comment by Esmolante** [comment]
By the way. This would also satisfy #3000: it would be trivial to implement an option for most twisty route possible for bikers and other asphaltic fauna. :)

**Comment by muralito** [comment]
There are/was some efforts to generate real speed data https://github.com/OpenStreetMapSpeeds

**Reply by Esmolante** [reply]
Interesting, thank you.
Calculating speed from real data (as Google does in real time) would be even better (bridging privacy concerns). But, until then, the calculation I propose would greatly improve route calculation and arrival time estimation. The idea is simple and intuitive: you can't expect to go at the same speed on a straight road as on a road full of sharp bends and a steep gradient, even if both roads are of the same category (which also matters, because of the width of the roadbed).

**Reply by muralito** [reply]
I agree. Also another factor that greatly determines the practical speed is the type of vehicle. E.g. in a unpaved country road, for a standard car maybe its safe to do 30-40km/h to not broke the car, but in the same road a light truck you can do 60-80 without risk. Maybe using the type of vehicle and the OSM "smoothness" key value (or assuming a good default value, or worldwide defaults and territory defaults by highway type) to adjust the practical speed.

**Reply by muralito** [reply]
Maybe as an option, to allow users the adjusted calculation, or the traditional one... or even something more parameterizable than on/off, like been able to set the weight/penalties of each condition.
The maxspeed settings is also that somehow varies between users. Some people does not exceed the maximum speed under no circunstances, but others users are not that strict.

**Reply by Esmolante** [reply]
Or, for example, a van, a motorhome or a lorry. Their maximum speed (in some countries) is slower than general vehicles. It would be easy, I think, modify calculation according to a parameter "type of vehicle" (or manually enter speed for streets, roads and highways).

There're two big problems with using maxspeed to calculate the route:
- winding roads 
- towns 
In that cases, maxspeed is far from representative, so the route calculation is poor (as well as the estimated time of arrival). The better example in my case is a calculated route crossing Madrid from end to end instead of using ring roads. It is not a poor route, it is absurd.

**Reply by biodranik** [reply]
Please, don't push. ETA will be more realistic when we properly add penalties for traffic lights and turns. These are ignored now, which leads to most of the mentioned issues.

**Reply by Esmolante** [reply]
I was not rushing in any way. 
I'm sorry you interpreted it that way, I didn't mean to. I just wanted to discuss aspects that I think could improve the programme. They will of course be addressed when the programmers see fit.
penalties for traffic lights and turns --> Traffic jams. They have a certain predictability, but I can't think how to introduce it in the calculation.
most of the mentioned issues --> Half of the issues. The other are curves and slopes. But in this case we do know the geometry of the road to evaluate it.
That's what I was proposing, with all due respect and consideration to those of you who are developing this project.

**Reply by muralito** [reply]
I wasnt rushing either.

You mention the traffic jams, as they introduces a problem with time averages. Some trip times have great variation depending if there are done in rush hours.

Also, some traffic lights may introduce some delay, but other not, specially in avenues, you just wait for the first light, and then go with the green wave, without being delayed in the next traffic lights.

Anyways, maybe some framework to log the time predicted and the real trip time is useful to improve the calculations, in particular if the estimated time is lower than the real.

**Reply by Esmolante** [reply]
traffic jams --> We can establish some regularities. For example, in each country, rush hour is at a certain time. Then, at that hour, we can add 100% more time on urban roads. One hour earlier, 70% more time (this is just an example). Thus, we add a coefficient according to the time of day which corrects the maxspeed.
traffic lights --> You are right, it can vary a lot, but it will always be better to introduce a time penalty in them than not to do so. Also at turns, as #biodranik says, or at pedestrian crossings.
Another parameter we could take into account is the number of lanes. It is more frequent that the phenomenon you say, of traffic lights turning on at the same time, occurs on multi-lane street roads (wide avenues).
None of this would be perfect, but it would give better time estimates and therefore better routes. And it would be a differentiating feature from other browsers (except Google, which plays in another division).

**Reply by Misalf-git** [reply]
Elaborated guessing is still guessing.

**Reply by Esmolante** [reply]
Some assumptions are more accurate than others.
Don't lose sight of the goal: to offer the fastest route, without real data.
Any ideas on how to improve the calculation?

**Reply by Misalf-git** [reply]
Taking real obstacles from OSM data or real-time traffic data into account is different from assuming the presence of traffic disturbances based on daytime, rush-hour, soccer finals, apprehension or moon phase.

**Reply by muralito** [reply]
> Elaborated guessing is still guessing.

Sure. The quality of a theory can be evaluated by comparing their predictions with the real events data.
A framework logging the predictions and the events data could be useful for real world experiments. E.g. some set of users using experimental calculations, while the control set uses the traditional stable ones, and the collected results data could be used for evaluation.
Maybe also taking some input/poll from the user when the trip finishes.

**Reply by Esmolante** [reply]
Time is correlated with traffic intensity. Moon phases do not. Please, let's be serious. 
Every parameter at our disposal that leads to increase the accuracy of the calculation and thus the quality of the calculated route should be incorporated. Anything that does not offer advantages should be disregarded.

**Reply by Esmolante** [reply]
E.g. some set of users using experimental calculations, while the control set uses the traditional stable ones, and the collected results data could be used for evaluation. --> Great idea. I offer myself for data collection. 
Another more convenient way would be to take advantage of the results provided by Google Maps routes (based on real data, let's remember) to establish patterns.
If the developers agree, I think I could be of help here.

**Reply by biodranik** [reply]
There are two independent tasks to improve OM ETA:
1. Take into account traffic lights, bumps, turns, road curves, and elevation.
2. Create, join or buy/get an online service that collects traffic-related data and road accidents/road works.

There are already some issues here for both points. Any help is appreciated.

**Reply by Esmolante** [reply]
Real-time traffic is undoubtedly the best solution, but it poses three problems:
a) Privacy
b) OM would no longer be offline-oriented
c) If provided by OM, it implies a critical mass of users, as well as a lot of resources to handle all that information.

The technical part I think would not be so difficult: OM collects and sends the average speed on each section of the road on which each user is traveling. If this deviates from the maxspeed (or, in the future, from the reference speed on each road, obtained with historical data), i.e. congestion, it sends an alert to the app of the users who have that road on their route to force a recalculation introducing that delay. Also for alerts of accidents, speedcams...
The problem would be the economic model to support it.

I am sorry if I am stating the obvious, I would like to be of more help.

**Reply by Misalf-git** [reply]
I'm not trying to talk you down, but you're killing your own arguments. With your approach, a, b and c still hold true.

**Reply by Esmolante** [reply]
I am just discussing the subject, pros and cons.

**Comment by muralito** [comment]
For the city where I live, there are open data with the average vehicle speed for each lane, measured every 5 minutes in the main streets.
One csv file per month since january 2021.
https://catalogodatos.gub.uy/dataset/intendencia-montevideo-velocidad-promedio-vehicular-en-las-principales-avenidas-de-montevideo

Let me know if such data is useful.

**Comment by Esmolante** [comment]
Not exactly the point but I don't want to risk unleashing the wrath of the programmers by opening new threads. :)
ETA is always calculated based on maxspeed, and does not even take into consideration the historical average speed on each type of road of each driver. For example, ETA is the same whether you drive a car or a motorhome or a lorry, even if the legal speed on the motorway is 120km/h with one and 90km/h with the other.
One option has already been discussed, which is that the user chooses the type of vehicle (or directly the maximum speeds for each type of road). 
The other would be for the programme itself to record the average speed and react accordingly. For example, if I have been driving for 3 hours on a motorway at 90km/h, it does not assume that I will be driving at 120km/h for the rest of the journey. 
Both options would increase the accuracy of the ETA, as well as the quality of the route (if you can't exceed 90km/h, you may not want to take a detour on the motorway, because you can't take advantage of its higher speed limit).

**Reply by Misalf-git** [reply]
> The other would be for the programme itself to record the average speed and react accordingly.

That's a good one.

**Reply by biodranik** [reply]
Good one, but it can't predict right in advance, before collecting the data.

**Comment by Esmolante** [comment]
On another note, and without wishing to monopolise the debate.
I was thinking of a feasible intermediate step to reach real-time traffic speed. 
To know this data we need a huge mass of connected users (Google again), so that there are several OM users on each (at least important) road/street at more or less the same time. This is still a long way off.
However, what would be more feasible is that, with the user's permission and properly anonymised, the programme could collect the average speed at which the driver is travelling on each road (it would be interesting to relate this to a vehicle class, car, motorbike, lorry, van / motorhome...). 
These average speeds of each road are sent to the server from time to time and, based on them, OM could gradually build up a database of realistic average speeds for each day, discriminating by day of the week and time of day. For this, a much smaller mass of users is sufficient, so that at least once in a while a user passes each road (at least the main roads). For roads without data, until they are available, the average speed could be obtained by extrapolating from similar roads (same category in the vicinity). This alone would be a huge improvement in the quality of the optimal route calculation.

**Reply by biodranik** [reply]
Let's focus here on everything we can do in offline except real time and historical traffic data. It worth to discuss its implementation under its own issue (I think there was one already).

**Reply by Esmolante** [reply]
All right, agreed. ;)

I will use the same example I used for the graphics issue #2970 

Let's take this route: 
![Screenshot_2022-10-18-14-19-08-306_app organicmaps](https://user-images.githubusercontent.com/109362590/196432415-5d5f80d0-dc1f-47ba-b619-8aa80d1160e7.jpg)
The algo chooses it because it is the shortest route on streets having the same maxspeed. But it is the wrong choice - no one in the area would think of taking this shortcut instead of following the main road, even if it is a few metres longer.

![Screenshot_2022-10-18-14-20-33-279_cz seznam mapy](https://user-images.githubusercontent.com/109362590/196440728-b3999271-255b-48dd-95f4-5a204acef490.jpg)


How to prevent such a mistake? 
a) Add a penalty of a few seconds for each detour. When entering another street, you have to slow down, stop if there is a stop sign, look, wait, enter the other lane and then speed up again. For a few metres, it's probably not worth the trouble.
b) Not assigning by default to each track its maxspeed, but a more realistic value. In this case, a highway=tertiary cannot be assumed to have the same average speed as a highway=residential, even if both are in urban areas. 

The main function of a navigator is to provide the best route between two points. In large part, this is why a user chooses one or the other. If we can get OM to deliver better results than the competition (better approximations, correct assumptions), it will succeed. ;)

**Reply by coyoteecd** [reply]
@Esmolante @biodranik Another option is to add a penalty based on road class of each segment. You could even make it a routing option such as "Prefer main roads". Road class is available in OSM as [highway](https://wiki.openstreetmap.org/wiki/Key:highway) tag. If the current algo chooses a route based on max speed of segments, I think it would be easier to add these penalties instead of analyzing detours.

In the picture above, the "main road" is classified as tertiary road, while the shortcut is classified as residential road. Adding a penalty for all those residential segments would make the algo stick to the main road since it would become the fastest route. Of course, at some point, even with penalty, the residential road would still get picked. In that case I would argue it is really a useful shortcut worth taking.

Thoughts?

**Reply by biodranik** [reply]
@coyoteecd how do you want to combine maxspeed with your class approach? Ignoring maxspeed is a bad idea.

**Reply by coyoteecd** [reply]
I was thinking of putting a penalty on the maxspeed, e.g. 
* highway == trunk road, use maxspeed * 1
* highway == primary road, use maxspeed * x
* highway == secondary road, use maxspeed * y
...
and so on, where 0 < x/y/z <= 1 and  would probably require some tweaking (they can be 1, for instance trunk vs primary might not require a penalty). For example use 0.9 for secondary roads and maybe a more aggressive value for residential ones.

This basically decreases the maxspeed based on the road class, as far as routing is concerned. It is not perfect, but as stated above I think it's simpler to try out and might get better results than purely using maxspeed (especially given that many villages/small towns have 50 km/h flat everywhere).

**Reply by biodranik** [reply]
We already have a not-perfect approach. Let's make it perfect ) Your idea doesn't work with speed bumps, traffic lights, and it will influence the otherwise valid ETA in some cases.

**Reply by muralito** [reply]
Maybe it could work, but not with the same penalties worldwide. What works for your city maybe is not the best for mine. The parameters in your approach may need to be tuned for each country/state/city.

**Reply by Esmolante** [reply]
We agree. The only thing is that these X and Y should be variable to add other parameters besides the category (road geometry, gradient, turns...).

**Reply by Esmolante** [reply]
> We already have a not-perfect approach. Let's make it perfect 

It is not possible to predict with total accuracy the speed on each road (and therefore the ETA), only to estimate it. It can be done in various ways (from more to less refined):
A) real-time data from other drivers on that road. We assume that the driver will go at the same speed as everyone else (which may be false, e.g. a lorry, and the speed would have to be corrected with the preceding speeds of that user).
B) historical data from other drivers on that road, discriminating day and time. We assume that the traffic pattern will repeat itself (quite likely, and is self-correcting with new data. Needs the same correction for driver/vehicle type as A)
C) maxpeed of the road. We assume that the driver will always drive at the maximum allowed speed of the road, which is a very poor approximation, and needs to be improved by taking into account, in addition to A) and B), other modifiers such as traffic lights, road geometry, detours...
The key is to establish assumptions that allow a more realistic calculation.
